### PR TITLE
Initial support for parsing JSON.

### PR DESCRIPTION
**Pull request by ryzhyk**
This is not yet complete, but I'd like to merge this to avoid future rebasing problems.

### CODE REVIEW

File: crates/adapters/src/format/json/input.rs [hunk 1]
```diff
@@ -0,0 +1,1207 @@
...
+/// Only the `payload` field is currently supported; other fields are ignored.
+#[derive(Debug, Deserialize)]
+pub struct DebeziumUpdate<'a> {
+    #[serde(borrow)]
+    payload: DebeziumPayload<'a>,
+}
+
+/// Schema of the `payload` field of a Debezium data change event.
+#[derive(Debug, Deserialize)]
+pub struct DebeziumPayload<'a> {
```
Comment (mihaibudiu) [inline crates/adapters/src/format/json/input.rs:181]: they should provide a library for this.
maybe this can be a separate crate?

**Review by mihaibudiu — APPROVED** [review]

File: crates/adapters/src/transport/mod.rs [hunk 2]
```diff
@@ -154,8 +152,18 @@ pub trait InputEndpoint: Send {
...
+    ///
+    /// `data` is not guaranteed to start or end on a record boundary.
+    /// The parser is responsible for identifying record boundaries and
+    /// buffering incomplete records to get prepended to the next
+    /// input fragment.
+    fn input_fragment(&mut self, data: &[u8]) -> AnyResult<()>;
+
     /// Push a chunk of data to the consumer.
-    fn input(&mut self, data: &[u8]) -> AnyResult<()>;
+    ///
```
Comment (gz) [inline crates/adapters/src/transport/mod.rs:164]: stray comment

File: crates/adapters/src/util.rs [hunk 3]
```diff
@@ -0,0 +1,12 @@
+/// Returns the index of the first character following the last newline
+/// in `data`.
+pub(crate) fn split_on_newline(data: &[u8]) -> usize {
```
[the remark this answers is not in the document]
Comment (gz) [reply crates/adapters/src/util.rs:3]: might be ok for code below

**Review by gz — APPROVED** [review]
didnt review java

File: crates/adapters/src/format/mod.rs [hunk 4]
```diff
@@ -49,19 +112,42 @@ impl dyn InputFormat {
...
     /// if parsing fails.
-    fn input(&mut self, data: &[u8]) -> AnyResult<usize>;
+    fn input_fragment(&mut self, data: &[u8]) -> AnyResult<usize>;
+
+    /// Push a chunk of data to the parser.
+    ///
+    /// The parser breaks `data` up into records and pushes these records
+    /// to the circuit using the
+    /// [`DeCollectionHandle`](`crate::DeCollectionHandle`) API.
+    /// The chunk is expected to contain complete records only.
```
[the remark this answers is not in the document]
Comment (ryzhyk) [reply crates/adapters/src/format/mod.rs:142]: The parser will return an error when trying to parse an incomplete record.

File: crates/adapters/src/format/mod.rs [hunk 5]
```diff
@@ -49,19 +112,42 @@ impl dyn InputFormat {
...
+    ///
+    /// The parser must not buffer any data.
```
[the remark this answers is not in the document]
Comment (ryzhyk) [reply crates/adapters/src/format/mod.rs:144]: I am trying to make sure that buffering happens in few well-defined placed in the pipeline.

File: crates/adapters/src/util.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (ryzhyk) [reply crates/adapters/src/util.rs:3]: I think it works for UTF-8, where all bytes of a multibyte character should star with `1`, so `\n` cannot be part of such a character.  Am I missing something else?

**Pull request merged**